react/src/components/Table/ColumnVisibilityModal.tsx

```diff
@@ -1,34 +1,74 @@
-import { useState } from 'react';
+import { useEffect, useState } from 'react';
 import { DragDropContext, Draggable, Droppable, DropResult } from 'react-beautiful-dnd';
 import { BsFillEyeFill, BsFillEyeSlashFill } from 'react-icons/bs';
 import { ColumnInstance, HeaderGroup, UseTableInstanceProps } from 'react-table';
-import { camelize } from '../../utils';
 import { Button, Checkbox, DragHandle, Flex, InlineFlex, Modal } from '../index';
 import { IconPadder } from './Table.styles';
 
 interface ColumnVisibilityModalProps<T extends object>
     extends Pick<UseTableInstanceProps<T>, 'toggleHideColumn'> {
     headerGroups: HeaderGroup<T>[];
-    toggleGroupVisibility: (g: HeaderGroup<T>) => void;
     allColumns: ColumnInstance<T>[];
+    visibleColumns: ColumnInstance<T>[];
     setColumnOrder: (update: string[] | ((columnOrder: string[]) => string[])) => void;
+    cached: Record<string, boolean>;
+    setCached: (value: Record<string, boolean>) => void;
 }
 
 export default function ColumnVisibilityModal<T extends {}>({
     headerGroups,
-    toggleGroupVisibility,
     toggleHideColumn,
+    cached,
+    setCached,
     allColumns,
+    visibleColumns,
     setColumnOrder,
 }: ColumnVisibilityModalProps<T>) {
     const [showModal, setShowModal] = useState<boolean>(false);
     const [order, setOrder] = useState<ColumnInstance<T>[]>([]);
+    // Cache a column's latest "check" status independent of the column's group status. Does not contain columns with type "empty".
+    const [cachedVisibility, setCachedVisibility] = useState<Record<string, boolean>>({});
+    useEffect(() => {
+        setCachedVisibility(cached);
+    }, [cached, showModal]);
+
+    // State to represent the current "check" status.
+    const [checkedColumns, setCheckedColumns] = useState<Record<string, boolean>>({});
+    useEffect(() => {
+        setCheckedColumns(
+            Object.fromEntries(
+                allColumns.filter(c => c.type !== 'fixed').map(c => [c.id, c.isVisible])
+            )
+        );
+    }, [allColumns, visibleColumns, showModal]);
+
+    const groupMapping: Record<string, string> = {
+        Variant: 'emptyCore',
+        'Variant Details': 'emptyVariationDetails',
+        'Case Details': 'emptyCaseDetails',
+    };
+
+    const isGroupExpanded = (header: string | undefined) => {
+        if (!header) return false;
+        return !checkedColumns[groupMapping[header]];
+    };
+
+    // Find a column's group: "Variant", "Variant Details" or "Case Details".
+    const columnToGroup = (columnId: string) => {
+        let group = headerGroups[0].headers[0];
+        headerGroups[0].headers.forEach(header => {
+            if (header.columns?.find(c => c.id === columnId)) group = header;
+        });
+        return group;
```

Comment:
Yes, since each column belongs to a group, the default group should not cause any issues. 

It's better to remove this default group though, and I just removed it.